### PR TITLE
[Trilinos] adding Amesos2 support

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
Adding support for the Amesos2 direct solvers

this is a WIP, I don't get why it is not linking correctly against the Amesos2 libs, even though they are found correctly

requires #7900

### CODE REVIEW

File: applications/TrilinosApplication/CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +28,14 @@ endif()
 include_directories(SYSTEM ${TRILINOS_INCLUDE_DIR})
 link_directories(${TRILINOS_LIBRARIES})
 
+if(${TRILINOS_EXCLUDE_AMESOS2_SOLVER} MATCHES ON)
+    add_definitions(-DTRILINOS_EXCLUDE_AMESOS2_SOLVER)
+    message("**** WARNING: Manually disabled Amesos2-Solver ****")
+elseif(DEFINED ${HAVE_AMESOS2})
```
Comment (philbucher) [inline applications/TrilinosApplication/CMakeLists.txt]: @roigcarlo I cannot get this to work, tried several different things
=> should work in combination with `HAVE_AMESOS2` from `FindTRILINOS.cmake`

the idea is to automatically disable it if Amesos2 is not part of the trilinos installation

Can you have a look please?

File: applications/TrilinosApplication/external_includes/amesos2_solver.h [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+        // it is not strictly necessary to use rcp, however skipping it could not yet be achieved
+        // in any case doesn't hurt to have it
+        using Teuchos::RCP;
+        using Teuchos::rcp;
+
+        // Amesos2 is missing specializations for FE-matrices and vectors
+        // hence here we need to use the corresponding baseclasses
+        // see "Amesos2_MatrixTraits.hpp" and "Amesos2_VectorTraits.hpp"
+        using MAT = Epetra_CrsMatrix;   // baseclass of "SparseMatrixType" aka "Epetra_FECrsMatrix"
+        using MV  = Epetra_MultiVector; // baseclass of "VectorType"       aka "Epetra_FEVector
```
Comment (philbucher) [inline applications/TrilinosApplication/external_includes/amesos2_solver.h:165]: For the record, I tried to provide manually the MatrixAdapter for the FE-Matrix & Vector, but was not successful (see https://docs.trilinos.org/dev/packages/amesos2/doc/html/classAmesos2_1_1MatrixAdapter.html)

in the future when we switch to Kratos Matrices I think we can take another look, we can probably make it work without copies by providing a suitable adapter

**Comment by philbucher** [comment]
seems like Amesos2 also requires TPetra => #7906

**Comment by philbucher** [comment]
nice, it also depends on Kokkos :DDD

**Comment by philbucher** [comment]
Summary for the record (and as note to myself for extensions in the future):
- Amesos2 requires the following other Trilinos Packages:
    - Kokkos
    - TPetra (not sure if headers are sufficient)
    - ShyLU (I think only fot he support of some solvers. If those are disabled then probably it is not needed)
    Those packages are automatically enabled when enabling Amesos2 support during compilation of Trilinos
- When enabling other solvers (like the new Basker solver) it seems to be necessary to also link against Kokkos, but not 100% sure. => Will be investigated in follow up PRs
- With the Matrix Adapters mentioned above I think it will be possible to directly use the new Kratos Csr Matrix => #6623 
- I added a mechanism that automatically disabled this solver (such that the compilation of the TrilinosApp doesn't crash if Amesos2 is not installed). I am planning to use this mechanism also for the other solvers (aka Amesos, ML, Aztec) such that we can keep the dependencies to Trilinos to an absolute minimum! => @sunethwarna this is what I asked you to try out :)

Also since the Trilinos Docs are often a bit outdated here is some interesting info I would:
- Amesos supports the following solvers:
    - KLU (natively) => pretty slow, see #7846 
    - [SuperLUDist](https://github.com/xiaoyeli/superlu_dist) up to version 2.5
    - [MUMPS](http://mumps.enseeiht.fr/): the latest version (5.x.x) is supported
        Useful Infos:
        - https://github.com/trilinos/Trilinos/issues/6339
        - https://www.code-aster.org/forum2/viewtopic.php?id=18500
    - others I am not sure, check [here](https://trilinos.github.io/amesos.html), although it seems to be outdated
- Amesos2 supports the following solvers:
    - KLU2 (natively, templated version of KLU)
    - [SuperLUDist](https://github.com/xiaoyeli/superlu_dist) up to version 5.4 (support for the latest version (6) is requested, see e.g. [here](https://github.com/trilinos/Trilinos/issues/6892))
    - [MUMPS](http://mumps.enseeiht.fr/): the latest version (5.x.x) is supported
    - Basker (natively, but needs to be enabled at compile time in Trilinos and seems to be dependent on other packages like ShyLU but not sure)
    - Recently support for the [STRUMPACK solver](https://github.com/pghysels/STRUMPACK) was [added](https://github.com/trilinos/Trilinos/pull/8192), looks promising
    - others I am not sure, check [here](https://trilinos.github.io/amesos2.html), although it seems to be outdated

Overall info of supported solvers: https://github.com/trilinos/Trilinos/issues/7708

**Comment by philbucher** [comment]
Note to self:
Also update the Readme of the TrilinosApp :)

**Review by roigcarlo — DISMISSED** [review]
Can't see anything wrong. I've tested also in my machine and seems to compile. GJ!!!

**Comment by philbucher** [comment]
now Amesos2 is disabled by default, as I still consider it under development

@sunethwarna could you please compile this branch in your machine to make sure it works?

**Comment by sunethwarna** [comment]
on it :)

**Comment by sunethwarna** [comment]
I compiled this branch with gcc 10.2, trilinos 13.0.1, and compiles. I ran full debug tests in few applications, they are working as well. (Those tests are not using Amesos btw ).

**Comment by philbucher** [comment]
Awesome thx!
Can you approve please?

**Review by sunethwarna — APPROVED** [review]
:)

**Pull request merged**